### PR TITLE
Add ServiceMatch support to Windows data plane

### PR DESCRIPTION
**Pull request by caseydavenport**
## Description

WIP

## Todos
- [ ] Unit tests (full coverage)
- [ ] Integration tests (delete as appropriate) In plan/Not needed/Done
- [ ] Documentation
- [ ] Backport
- [ ] Release note

## Release Note

```release-note
None required
```

### CODE REVIEW

File: dataplane/windows/policysets/policysets.go [hunk 1]
```diff
@@ -361,6 +361,71 @@ func (s *PolicySets) protoRuleToHnsRules(policyId string, pRule *proto.Rule, isI
...
+
+		// We need to consolidate addrs based on ports/protocols, so use a map to lookup.
+		membersByPort := map[string]*members{}
+
+		// We need to ensure the ordering of generated rules is deterministic, so use a slice.
+		orderedMembers := []*members{}
+		for _, m := range ipsetMembers {
+			// The member should be of the format <IP>,(tcp|udp):<port number>
+			addr, proto, port := parseIPPortMember(m)
+			var m *members
```
Comment (song-jiang) [inline dataplane/windows/policysets/policysets.go:402]: m is used as the loop variable so we can declare `var pm *policyMember` here

File: dataplane/windows/policysets/policysets.go [hunk 2]
```diff
@@ -361,6 +361,71 @@ func (s *PolicySets) protoRuleToHnsRules(policyId string, pRule *proto.Rule, isI
...
+			proto uint16
+			port  string
+			addrs []string
+		}
+
+		// We need to consolidate addrs based on ports/protocols, so use a map to lookup.
+		membersByPort := map[string]*members{}
+
+		// We need to ensure the ordering of generated rules is deterministic, so use a slice.
+		orderedMembers := []*members{}
```
Comment (song-jiang) [inline dataplane/windows/policysets/policysets.go:398]: orderedPolicyMember?

File: dataplane/windows/policysets/policysets.go [hunk 3]
```diff
@@ -361,6 +361,71 @@ func (s *PolicySets) protoRuleToHnsRules(policyId string, pRule *proto.Rule, isI
 		logCxt.WithField("action", ruleCopy.Action).Panic("Unknown rule action")
 	}
 
+	//
+	// DstIpPort sets - these cannot co-exist with other fields, so do them first and short-circuit if set.
+	//
+	if len(ruleCopy.DstIpPortSetIds) > 0 {
+		ipsetMembers, err := s.getIPSetAddresses(ruleCopy.DstIpPortSetIds)
+		if err != nil {
+			logCxt.Warn("DstIpPortSetIds could not be resolved, rule will be skipped")
```
Comment (song-jiang) [inline dataplane/windows/policysets/policysets.go:370]: If a policy reference a service but all backend endpoints are degraded, will we reach here?

File: dataplane/windows/policysets/policysets_test.go [hunk 4]
```diff
@@ -376,6 +376,116 @@ func TestRuleRendering(t *testing.T) {
...
+			RemoteAddresses: "10.0.0.2",
+			RemotePorts:     "80",
+		},
+		// Default deny rule.
+		{Type: hns.ACL, Protocol: 256, Action: hns.Block, Direction: hns.Out, RuleType: hns.Switch, Priority: 1001},
+		// Default host/pod rule.
+		{Type: hns.ACL, Protocol: 256, Action: hns.Allow, Direction: hns.Out, RuleType: hns.Host},
+	}), "unexpected rules returned for IP+port policy")
+}
+
```
Comment (song-jiang) [inline dataplane/windows/policysets/policysets_test.go:488]: Suggest to add a test case for empty ipset.

**Review by song-jiang — COMMENTED** [review]
Looking good with a few comments.

File: dataplane/windows/policysets/policysets.go (hunk 3 above)
Comment (caseydavenport) [reply dataplane/windows/policysets/policysets.go:370]: Nope, it will return `[]string{}, nil` and we'll properly generate just the default deny rules.

I added a UT for this case according to your recommendation :)

**Comment by caseydavenport** [comment]
@song-jiang I think this is ready for another look - added new UTs and an FV test and both are passing.

File: fv/winfv/policy_test.go [hunk 5]
```diff
@@ -117,5 +131,34 @@ var _ = Describe("Windows policy test", func() {
 			err := kubectlExec(`-t porter -- powershell -Command 'Invoke-WebRequest -UseBasicParsing -TimeoutSec 5 www.google.com'`)
 			Expect(err).To(HaveOccurred())
 		})
+		It("porter pod can connect to kube apiserver after creating service egress policy", func() {
+			// Assert API is not reachable.
+			err := kubectlExec(`-t porter -- powershell -Command 'Invoke-WebRequest -UseBasicParsing -TimeoutSec 5 https://kubernetes.default'`)
```
Comment (song-jiang) [inline fv/winfv/policy_test.go:136]: I think you should use `https://kubernetes.default.svc.cluster.local` as target (same as the connection made below). Because if there is an DNS issue with short format of service name, the test can still pass with or without applying service policy.

**Comment by song-jiang** [comment]
One change required but I don't think I need to review again. LGTM

**Review by song-jiang — APPROVED** [review]
